components/renku_data_services/data_connectors/db.py

```diff
@@ -0,0 +1,225 @@
+                )
+            return data_connector.dump()
+
+    @with_db_transaction
+    async def insert_data_connector(
+        self,
+        user: base_models.APIUser,
+        data_connector: models.UnsavedDataConnector,
+        *,
+        session: AsyncSession | None = None,
+    ) -> models.DataConnector:
+        """Insert a new data connector entry."""
+        if not session:
+            raise errors.ProgrammingError(message="A database session is required.")
+        ns = await session.scalar(
+            select(ns_schemas.NamespaceORM).where(ns_schemas.NamespaceORM.slug == data_connector.namespace.lower())
+        )
+        if not ns:
+            raise errors.MissingResourceError(
+                message=f"The data connector cannot be created because the namespace {
+                    data_connector.namespace} does not exist."
+            )
+        if not ns.group_id and not ns.user_id:
+            raise errors.ProgrammingError(message="Found a namespace that has no group or user associated with it.")
+
+        if user.id is None:
+            raise errors.UnauthorizedError(message="You do not have the required permissions for this operation.")
+
+        slug = data_connector.slug or base_models.Slug.from_name(data_connector.name).value
+
+        existing_slug = await session.scalar(
+            select(ns_schemas.EntitySlugORM)
+            .where(ns_schemas.EntitySlugORM.namespace_id == ns.id)
+            .where(ns_schemas.EntitySlugORM.slug == slug)
+        )
+        if existing_slug is not None:
+            raise errors.ConflictError(message=f"An entity with the slug '{ns.slug}/{slug}' already exists.")
+
+        visibility_orm = (
+            apispec.Visibility(data_connector.visibility)
+            if isinstance(data_connector.visibility, str)
+            else apispec.Visibility(data_connector.visibility.value)
+        )
+        data_connector_orm = schemas.DataConnectorORM(
+            name=data_connector.name,
+            visibility=visibility_orm,
+            storage_type=data_connector.storage.storage_type,
+            configuration=data_connector.storage.configuration,
+            source_path=data_connector.storage.source_path,
+            target_path=data_connector.storage.target_path,
+            readonly=data_connector.storage.readonly,
+            created_by_id=user.id,
+            description=data_connector.description,
+            keywords=data_connector.keywords,
+        )
+        data_connector_slug = ns_schemas.EntitySlugORM.create_data_connector_slug(
+            slug, data_connector_id=data_connector_orm.id, namespace_id=ns.id
+        )
+
+        session.add(data_connector_orm)
+        await session.flush()
+        session.add(data_connector_slug)
+        await session.flush()
+        await session.refresh(data_connector_orm)
```

Comment:
question: I overlooked this change in the project db in the previous pr https://github.com/SwissDataScienceCenter/renku-data-services/pull/384 but I thought you could do a single flush and the orm would take care of doing it in the correct order (as long as the foreign key relationships are the right way around)? Did having it with a single flush break something?

Reply:
Yes, doing it in a single flush saves the `data_connector_slug` first for some reason, which surfaces as a foreign key constraint violation. My intuition on this is that alembic may not guess the correct order because our `ORM` classes are split across different bases. Alembic was issuing the correct order when the two tables where in the same `schema` and the `ORM` classes derived from the same base.

One potential solution may be to enforce using a single `ORM` base (though it may be annoying to define different schemas for different tables).

Reply:
well if it causes issues then that's fine, having to talk to the db twice isn't a big enough downside to worry about it. Though we might want to explain this in a comment in code.

Reply:
actually I think it might be due to using `relationship()` both on `EntitySlugORM`  and `DataConnectorORM`, without a `back_populates` on the `DataConnectorORM`. That might screw up the sqlalchemy depencency checker (as in, it just sees two unrelated `relationship`'s)
from the [docs](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship.params.back_populates):
> This allows objects on both sides of each [relationship()](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship) to synchronize in-Python state changes and also provides directives to the [unit of work](https://docs.sqlalchemy.org/en/20/glossary.html#term-unit-of-work) flush process how changes along these relationships should be persisted.

Reply:
OK, so this was a big head-scratcher because `back_populates` is defined on both sides. The issue was that `viewonly=True` deactivates persistence related features, so `viewonly=True` had to be removed from the slug ORM to make this work.

Reply:
Done in 2e4f53b7c156c74455c476d79380ddcc837e429f.